rclcpp/include/rclcpp/strategies/message_pool_memory_strategy.hpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+#define RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+
+#include <rclcpp/macros.hpp>
+#include <rclcpp/message_memory_strategy.hpp>
+
+namespace rclcpp
+{
+namespace strategies
+{
+namespace message_pool_memory_strategy
+{
+
+template<typename MessageT>
+struct PoolMember
+{
+  std::shared_ptr<MessageT> msg_ptr_;
+  bool used;
+};
+
+template<typename MessageT, size_t size,
+typename std::enable_if<rosidl_generator_traits::has_fixed_size<MessageT>::value>::type * =
+nullptr>
+class MessagePoolMemoryStrategy
+  : public message_memory_strategy::MessageMemoryStrategy<MessageT>
+{
+public:
+  RCLCPP_MAKE_SHARED_DEFINITIONS(MessagePoolMemoryStrategy);
+  MessagePoolMemoryStrategy()
+  : next_array_index_(0)
+  {
+    for (size_t i = 0; i < size; ++i) {
+      pool[i].msg_ptr_ = std::make_shared<MessageT>();
+      pool[i].used = false;
+    }
+  }
+
+  std::shared_ptr<MessageT> borrow_message()
+  {
+    size_t current_index = next_array_index_;
+    next_array_index_ = (next_array_index_ + 1) % size;
+    if (pool[current_index].used) {
+      throw std::runtime_error("Tried to access message that was still in use! Abort.");
+    }
+    pool[current_index].msg_ptr_->~MessageT();
+    new (pool[current_index].msg_ptr_.get())MessageT;
```

Comment:
Does this compile? I've always called the constructor explicitly, it would be like this:

```c++
new (pool[current_index].msg_ptr_.get()) MessageT();
```

(note the trailing `()`) If this compiles, then no worries, I just didn't know you could do that.